auth-source-pass.el

```diff
@@ -279,24 +277,16 @@ Entries matching supported formats that combine HOSTNAME, USER, &
 PORT are accumulated into sublists where the car of each sublist
 is a regular expression for matching paths in the password-store
 and the remainder is the list of matching entries."
-  (let ((suffix-match-lists
-         (mapcar (lambda (suffix) (list (format "\\(^\\|/\\)%s$" suffix)))
+  (let ((suffixes
+         (mapcar (lambda (suffix) (format "\\(^\\|/\\)%s$" suffix))
                  (auth-source-pass--generate-entry-suffixes hostname user port))))
-    (cl-reduce #'auth-source-pass--entry-reducer
-               (auth-source-pass-entries)
-               :initial-value suffix-match-lists)))
-
-(defun auth-source-pass--entry-reducer (match-lists entry)
-  "Match MATCH-LISTS sublists against ENTRY.
-
-The result is a copy of match-lists with the entry added to the
-end of any sublists for which the regular expression at the head
-of the list matches the entry name."
-  (mapcar (lambda (match-list)
-            (if (string-match (car match-list) entry)
-                (append match-list (list entry))
-              match-list))
-          match-lists))
+    (mapcar #'auth-source-pass--entries-matching-suffix suffixes)))
+
+(defun auth-source-pass--entries-matching-suffix (suffix)
+  "Return the list of entries matching SUFFIX."
+  (cl-remove-if-not
+   (lambda (entry) (string-match-p suffix entry))
+   (auth-source-pass-entries)))
```

Comment:
This implementation is simpler but I think it does have one disadvantage, which is that it will re-traverse the entire filename hierarchy in the password-store once for every suffix matcher, which could be up to 6 times. In a typical password-store of a few tens to a few hundreds of entries this is unlikely to be noticeable, but most of the complexity of my implementation was trying to avoid that.

I don't think this will cause me a problem in any of my use cases, so I'm not opposed to the change, but I would prefer a solution that only traversed the the password-store file hierarchy once. Alternatively, perhaps we could at least memorize the password-store filename traversal in a let here in `auth-source-pass--entries-matching-suffix so we're just going over the list multiple times instead of traversing the filesystem itself?

Reply:
Good catch. I followed your suggestion.